src/Microsoft.AspNetCore.Razor/CodeGenerators/Visitors/CSharpTagHelperFieldDeclarationVisitor.cs

```diff
@@ -86,10 +91,87 @@ protected override void Visit(TagHelperChunk chunk)
                 }
             }
 
+            if (!Context.Host.DesignTimeMode)
+            {
+                PreAllocateUnboundTagHelperAttributes(chunk);
+            }
+
             // We need to dive deeper to ensure we pick up any nested tag helpers.
             Accept(chunk.Children);
         }
 
+        private void PreAllocateUnboundTagHelperAttributes(TagHelperChunk chunk)
+        {
+            var boundAttributes = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+            for (var i = 0; i < chunk.Attributes.Count; i++)
+            {
+                var attribute = chunk.Attributes[i];
+                var associatedDescriptors = chunk.Descriptors.Where(descriptor =>
+                    descriptor.Attributes.Any(attributeDescriptor => attributeDescriptor.IsNameMatch(attribute.Key)));
+
+                // If there's no descriptors associated or we're hitting a bound attribute a second time.
+                if (!associatedDescriptors.Any() || !boundAttributes.Add(attribute.Key))
+                {
+                    string preAllocatedAttributeVariableName = null;
+
+                    if (attribute.Value == null)
+                    {
+                        // For minimized attributes we add the null character to the end of the key to differentiate
+                        // between name and name=""; null characters are invalid in attribute values.
+                        var preAllocatedAttributeKey = attribute.Key + '\0';
+                        if (TryCachePreallocatedVariableName(preAllocatedAttributeKey, out preAllocatedAttributeVariableName))
+                        {
+                            Writer
+                                .Write("private static readonly global::")
+                                .Write(_tagHelperContext.TagHelperAttributeTypeName)
+                                .Write(" ")
+                                .Write(preAllocatedAttributeVariableName)
+                                .Write(" = ")
+                                .WriteStartNewObject(_tagHelperContext.TagHelperAttributeTypeName)
+                                .WriteStringLiteral(attribute.Key)
+                                .WriteEndMethodInvocation();
+                        }
+                    }
+                    else
+                    {
+                        string plainText;
+                        var isPlainText = CSharpTagHelperCodeRenderer.TryGetPlainTextValue(attribute.Value, out plainText);
+
+                        if (isPlainText)
+                        {
+                            var preAllocatedAttributeKey = attribute.Key + plainText;
```

Comment:
Is there a risk of key collisions here? (attribute.Key + plainText == attribute.Key for a bound attribute?)